File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/C._DFS/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
#include <algorithm>



void dfs(int& source, std::unordered_map<int, std::vector<int>>& adjList, int& v) {
    std::vector<std::string> colors (v + 1, "white");
    std::vector<int> callStack = {source};
    while (!callStack.empty()) {
        int& curVertex = callStack[callStack.size() - 1];
        callStack.pop_back();
        if (colors[curVertex] == "white") {
            std::cout << curVertex << " ";
            colors[curVertex] = "grey";
            callStack.push_back(curVertex);
            for (int& nextVertex : adjList[curVertex]) {
                if (colors[nextVertex] == "white") callStack.push_back(nextVertex);
            }
        } else if (colors[curVertex] == "grey") colors[curVertex] = "black";
    }
}


void solution() {
    /*
    Time Complexity O(ElogE + (V + E))
    Memory Complexity O(V + E)
    */
    int v, e;
    std::cin >> v >> e;
    std::unordered_map<int, std::vector<int>> adjList;
    for (int i = 0; i < e; ++i) {
        int source, destination;
        std::cin >> source >> destination;
        adjList[source].push_back(destination);
        adjList[destination].push_back(source);
    }
    int start;
    std::cin >> start;
    for (auto& pair : adjList) {
        std::sort(pair.second.begin(), pair.second.end(), std::greater<int>());
    }
    dfs(start, adjList, v);
}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/C._DFS/solution.py
import sys


def dfs(source: int, adj_list: dict[int, list[int]], v: int) -> None:
    colors: list[str] = ['white'] * (v + 1)
    call_stack: list[int] = [source]
    while call_stack:
        cur_vertex = call_stack.pop()
        if colors[cur_vertex] == 'white':
            sys.stdout.write(str(cur_vertex) + ' ')
            colors[cur_vertex] = 'grey'
            call_stack.append(cur_vertex)
            for next_vertex in adj_list.get(cur_vertex, []):
                if colors[next_vertex] == 'white':
                    call_stack.append(next_vertex)
        elif colors[cur_vertex] == 'grey':
            colors[cur_vertex] = 'black'



def solution() -> None:
    """
    Time Complexity O(ElogE + (V + E))
    Memory Complexity O(V + E)
    """
    v, e = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, list[int]] = dict()
    for _ in range(e):
        source, destination = map(int, sys.stdin.readline().rstrip().split())
        if source not in adj_list: adj_list[source] = []
        adj_list[source].append(destination)
        if destination not in adj_list: adj_list[destination] = []
        adj_list[destination].append(source)
    for vertex, edges in adj_list.items():
        edges.sort(reverse=True)
    start: int = int(sys.stdin.readline().rstrip())
    dfs(start, adj_list, v)


if __name__ == '__main__':
    solution()